src/hotspot/cpu/arm/sharedRuntime_arm.cpp

```diff
@@ -628,10 +627,9 @@ AdapterHandlerEntry* SharedRuntime::generate_i2c2i_adapters(MacroAssembler *masm
   const Register holder_klass   = Rtemp; // XXX should be OK for C2 but not 100% sure
   const Register receiver_klass = R4;
 
-  __ load_klass(receiver_klass, receiver);
-  __ ldr(holder_klass, Address(Ricklass, CompiledICHolder::holder_klass_offset()));
-  __ ldr(Rmethod, Address(Ricklass, CompiledICHolder::holder_metadata_offset()));
-  __ cmp(receiver_klass, holder_klass);
+  __ ic_check(1 /* end_alignment */);
+  __ ldr(Rmethod, Address(receiver_klass, CompiledICData::speculated_method_offset()));
```

Comment:
Maybe I am missing something here but this looks very wrong. The speculated `Klass*` gets loaded into `R4` (which `receiver_klass` alias) in `ic_check` this load would result in loading a `InstanceKlass*` c++ vtable pointer. 
`Ricklass` (`R8` alias) contains the  `CompiledICData*` .

I would think the correct diff would be
```diff
-  const Register receiver_klass = R4;
-
-  __ load_klass(receiver_klass, receiver);
-  __ ldr(holder_klass, Address(Ricklass, CompiledICHolder::holder_klass_offset()));
-  __ ldr(Rmethod, Address(Ricklass, CompiledICHolder::holder_metadata_offset()));
-  __ cmp(receiver_klass, holder_klass);
+  __ ic_check(1 /* end_alignment */);
+  __ ldr(Rmethod, Address(Ricklass, CompiledICData::speculated_method_offset()));
```

The fact that you say ARM32 tests are passing makes me doubt my understanding of the inline cache.

Reply:
hotspot jtreg tests pass in both variants, but you finding is correct. Updated the code with your suggestion.